pkg/nfspv/nfsvolume.go

```diff
@@ -0,0 +1,340 @@
+//		 timestamp for create event we are mutating deletion timestamp
+//		 fields with nil value.
+func (n *nfsVolume) CollectCreateEvents() (string, error) {
+
+	volumeData, err := n.getVolumeData()
+	if err != nil {
+		return "", err
+	}
+
+	if !n.isSupportedDataType() {
+		return "", errors.Errorf("data type %q is not supported. Supported types %v", n.dataType, supportedDataTypes)
+	}
+
+	if volumeData.NFSPVC != nil {
+		volumeData.NFSPVC.DeletionTimestamp = nil
+		volumeData.NFSPVC.DeletionGracePeriodSeconds = nil
+	}
+
+	volumeData.NFSPV.DeletionTimestamp = nil
+	volumeData.NFSPV.DeletionGracePeriodSeconds = nil
+
+	volumeData.BackingPVC.DeletionTimestamp = nil
+	volumeData.BackingPVC.DeletionGracePeriodSeconds = nil
+
+	volumeData.BackingPV.DeletionTimestamp = nil
+	volumeData.BackingPV.DeletionGracePeriodSeconds = nil
+
+	createData := &NFSCreateVolumeData{
+		VolumeProvisioned: volumeData,
+	}
+	rawData, err := json.Marshal(createData)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to marshal create volume events")
+	}
+	return string(rawData), nil
+}
+
+func (n *nfsVolume) CollectDeleteEvents() (string, error) {
+	// TODO: Is below check required?
+	if n.pvObj.DeletionTimestamp == nil {
+		return "", errors.Errorf("volume is not yet marked for deletion")
+	}
+	if !n.isSupportedDataType() {
+		return "", errors.Errorf("data type %q is not supported. Supported types %v", n.dataType, supportedDataTypes)
+	}
+
+	volumeData, err := n.getVolumeData()
+	if err != nil {
+		return "", err
+	}
+
+	createData := &NFSDeleteVolumeData{
+		VolumeDeleted: volumeData,
+	}
+	rawData, err := json.Marshal(createData)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to marshal delete volume events")
+	}
+	return string(rawData), nil
+}
+
+func (n *nfsVolume) AnnotateCreateEvent(pvObj *corev1.PersistentVolume) (*corev1.PersistentVolume, error) {
+	if pvObj.Annotations == nil {
+		pvObj.Annotations = make(map[string]string)
+	}
+	annoKey := n.annotationPrefix + collectorinterface.VolumeCreateEventAnnotation
+	pvObj.Annotations[annoKey] = collectorinterface.OpenebsEventSentAnnotationValue
+	return n.clientset.CoreV1().PersistentVolumes().Update(context.TODO(), pvObj, metav1.UpdateOptions{})
+}
+
+func (n *nfsVolume) AnnotateDeleteEvent(pvObj *corev1.PersistentVolume) (*corev1.PersistentVolume, error) {
+	pvCopy := pvObj.DeepCopy()
+	if pvCopy.Annotations == nil {
+		pvCopy.Annotations = make(map[string]string)
+	}
+	annoKey := n.annotationPrefix + collectorinterface.VolumeDeleteEventAnnotation
+	pvCopy.Annotations[annoKey] = collectorinterface.OpenebsEventSentAnnotationValue
+
+	patchBytes, _, err := getPatchData(pvObj, pvCopy)
+	if err != nil {
+		return nil, err
+	}
+	newPVObj, err := n.clientset.CoreV1().
+		PersistentVolumes().
+		Patch(context.TODO(), pvCopy.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Update inmemory reference
```

Comment:
why are we updating in-memory PV reference?

Reply:
Since we triggered the patch on PV for the next update it will return conflict errors. To avoid conflicting errors we are updating inmemory reference.

Reply:
ok